CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,475 @@
+goal of hardening passwords for use as encryption and authorization keys.
+
+## Abstract
+
+<!-- TODO(victor): Once we have a roughly agreed upon proposal for the pOPRF, link it here -->
+Given the required cryptographic functionality (to be discussed in more detail in a future
+proposal), this proposal describes an interface to be exposed by the ODIS service allowing the user
+to specify a domain string (i.e. a salt value) with rate-limiting rules to be combined with their
+blinded input (e.g.  phone number or password).
+
+Central to the proposed interface are domains. A domain is a struct which specifies rate limiting
+information and uniquely identifies the context of the secret to be hashed. As an example, a domain
+for hashing an account password might specify the username (context) and a cap of 10 password
+attempts (rate-limiting parameter). These would be combined with their password to form the final
+output.
+
+Queries with distinct domain specifiers will receive uncorrelated output. For example, output from
+ODIS with a phone number domain and message `+18002738255` will be distinct from and
+unrelated to the output when requesting with a password domain and message `+18002738255`.
+
+Additionally, domain specifiers will control what rate limiting rules are applied. For example the
+phone number domain could use the current ODIS rate limiting rules, while the password domain may
+use much stricter rules. Combined with the fact that distinct domains produce uncorrelated output,
+this allows for the domains to be used as distinct OPRFs with custom rate limiting rules. 
+
+In order to make this scheme flexible, allowing for user-defined tuning of rate-limits and the
+introduction of new rate limiting and authorization rules in the future, domains are defined as
+serializeable structs.
+
+## Motivation
+
+Effective [offline password cracking](https://en.wikipedia.org/wiki/Password_cracking) techniques
+limit the use of passwords to derive encryption or authentication keys. Rate-limited hashing can be
+used to make it much more difficult to crack a password. Computationally expensive password hashing
+functions, such as `PBKDF` and `scrypt`, are commonly used for this purpose, but provide [limited
+protection](https://arxiv.org/abs/2006.05023) and are expensive to run on end-user devices. ODIS,
+the phone number privacy service on Celo, implements a rate-limited hash to protect users' phone
+numbers from being extracted from their on-chain identifiers. With some extensions, ODIS can
+additionally be used to provide a rate-limited hash for passwords, as well as many more use cases.
+
+[Phone Number Privacy](https://docs.celo.org/celo-codebase/protocol/identity/phone-number-privacy)
+
+ODIS' rate-limited hash function is implemented from an
+[oblivious](http://www.cs.columbia.edu/~tal/4261/F19/hugo-columbia-oprf.pdf) [pseudorandom
+function](https://en.wikipedia.org/wiki/Pseudorandom_function_family) (OPRF). In effect, this allows
```

Comment:
To clarify, the rate limiting is independent from the hash function, right? As in the OPRF doesn't inherently limit how frequently an input can be hashed, it just ensures that the input is blinded while being hashed?

Reply:
Clarified below - rate limiting is part of the HTTP layer, not the OPRF crypto primitive.

Reply:
Right, but I might clarify further that HTTP is simply the transport (gRPC, or WebSockets could be used instead), so it's really at the "ODIS protocol" layer.